src/types.rs

```diff
@@ -136,3 +136,86 @@ impl LintPass for UnitCmp {
         }
     }
 }
+
+pub struct CastPass;
+
+declare_lint!(pub CAST_PRECISION_LOSS, Allow,
+              "casts that cause loss of precision, e.g `x as f32` where `x: u64`");
+declare_lint!(pub CAST_SIGN_LOSS, Allow,
+              "casts from signed types to unsigned types, e.g `x as u32` where `x: i32`");
+declare_lint!(pub CAST_POSSIBLE_TRUNCATION, Allow,
+              "casts that may cause truncation of the value, e.g `x as u8` where `x: u32`, or `x as i32` where `x: f32`");
+
+impl LintPass for CastPass {
+    fn get_lints(&self) -> LintArray {
+        lint_array!(CAST_PRECISION_LOSS,
+                    CAST_SIGN_LOSS,
+                    CAST_POSSIBLE_TRUNCATION)
+    }
+
+    fn check_expr(&mut self, cx: &Context, expr: &Expr) {
+        if let ExprCast(ref ex, _) = expr.node {
+            let (cast_from, cast_to) = (cx.tcx.expr_ty(&*ex), cx.tcx.expr_ty(expr));
+            if cast_from.is_numeric() && cast_to.is_numeric() && !in_external_macro(cx, expr.span) {
+                match (cast_from.is_integral(), cast_to.is_integral()) {
+                    (true, false) => {
+                        let from_nbits = match &cast_from.sty {
+                            &ty::TyInt(i) =>  4 << (i as usize),
```

Comment:
Correct, I thought of it after the fact. The default arm of the match also somewhat bothers me, because although it should never be reached (is_integral is only true for ty::IntTy and ty::UintTy according to rustc source), if it ever was, then it would successfully pass the `if from_nbits != 4` and cause shenanigans.

Like I said, it shouldn't happen but I'm considering putting 4 as the default match value to prevent it, or check and ensure that `from_nbits != 0`. Anyway, I will document it as what happens there is nowhere near obvious.